velox/common/base/VeloxException.h

```diff
@@ -55,8 +55,25 @@ struct ExceptionContext {
   /// Calls `messageFunc(arg)` and returns the result. Returns empty string if
   /// `messageFunc` is null.
   std::string message() {
-    return messageFunc ? messageFunc(arg) : "";
+    if (!messageFunc || suspended) {
+      return "";
+    }
+
+    std::string theMessage;
+
+    try {
+      // Make sure not to call messageFunc again in case it throws.
+      suspended = true;
+      theMessage = messageFunc(arg);
+      suspended = false;
+    } catch (...) {
+      return "Failed to produce additional context.";
+    }
+
+    return theMessage;
   }
+
+  bool suspended{false};
```

Comment:
It seems that the exception has hierarchy so I am not sure if it is better to use atomic here since we want to turn on tsan to avoid any potential data race.

Reply:
That's a good point. I ran the test with TSAN. There were no failures.

```
buck test @mode/dev-tsan velox/common/base/... -- ExceptionTest
```